atproto/identity/cache_catalog.go

```diff
@@ -0,0 +1,209 @@
+)
+
+type CacheCatalog struct {
+	HitTTL        time.Duration
+	ErrTTL        time.Duration
+	Inner         Catalog
+	mutex         sync.RWMutex
+	handleCache   map[syntax.Handle]HandleEntry
+	identityCache map[syntax.DID]IdentityEntry
+}
+
+type HandleEntry struct {
+	Updated time.Time
+	DID     syntax.DID
+	Err     error
+}
+
+type IdentityEntry struct {
+	Updated  time.Time
+	Identity *Identity
+	Err      error
+}
+
+var _ Catalog = (*CacheCatalog)(nil)
+
+func NewCacheCatalog(inner Catalog) CacheCatalog {
+	// TODO: these are kind of arbitrary default values...
+	hitTTL, err := time.ParseDuration("1h")
+	if err != nil {
+		panic(err)
+	}
+	errTTL, err := time.ParseDuration("2m")
+	if err != nil {
+		panic(err)
+	}
+	return CacheCatalog{
+		HitTTL:        hitTTL,
+		ErrTTL:        errTTL,
+		Inner:         inner,
+		handleCache:   make(map[syntax.Handle]HandleEntry, 10),
+		identityCache: make(map[syntax.DID]IdentityEntry, 10),
+	}
+}
+
+func (c *CacheCatalog) IsHandleStale(e *HandleEntry) bool {
+	if e.Err == nil && time.Since(e.Updated) > c.HitTTL {
+		return true
+	}
+	if e.Err != nil && time.Since(e.Updated) > c.ErrTTL {
+		return true
+	}
+	return false
+}
+
+func (c *CacheCatalog) IsIdentityStale(e *IdentityEntry) bool {
+	if e.Err == nil && time.Since(e.Updated) > c.HitTTL {
+		return true
+	}
+	if e.Err != nil && time.Since(e.Updated) > c.ErrTTL {
+		return true
+	}
+	return false
+}
+
+func (c *CacheCatalog) updateHandle(ctx context.Context, h syntax.Handle) (*HandleEntry, error) {
+	ident, err := c.Inner.LookupHandle(ctx, h)
+	if err != nil {
+		he := HandleEntry{
+			Updated: time.Now(),
+			DID:     "",
+			Err:     err,
+		}
+		c.mutex.Lock()
+		c.handleCache[h] = he
+		c.mutex.Unlock()
+		return &he, nil
+	}
+
+	entry := IdentityEntry{
+		Updated:  time.Now(),
+		Identity: ident,
+		Err:      nil,
+	}
+	he := HandleEntry{
+		Updated: time.Now(),
+		DID:     ident.DID,
+		Err:     nil,
+	}
+
+	c.mutex.Lock()
+	c.identityCache[ident.DID] = entry
+	c.handleCache[ident.Handle] = he
+	c.mutex.Unlock()
+	return &he, nil
+}
+
+func (c *CacheCatalog) ResolveHandle(ctx context.Context, h syntax.Handle) (syntax.DID, error) {
+	var err error
+	var entry *HandleEntry
+	c.mutex.RLock()
+	eObj, ok := c.handleCache[h]
+	c.mutex.RUnlock()
+
+	if !ok {
+		entry, err = c.updateHandle(ctx, h)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		entry = &eObj
+	}
+	if c.IsHandleStale(entry) {
+		entry, err = c.updateHandle(ctx, h)
+		if err != nil {
+			return "", err
+		}
+	}
+	return entry.DID, entry.Err
+}
+
+func (c *CacheCatalog) updateDID(ctx context.Context, did syntax.DID) (*IdentityEntry, error) {
+	ident, err := c.Inner.LookupDID(ctx, did)
+	entry := IdentityEntry{
+		Updated:  time.Now(),
+		Identity: ident,
+		Err:      err,
+	}
+	var he *HandleEntry
+	if err == nil && !ident.Handle.IsInvalidHandle() {
+		he = &HandleEntry{
+			Updated: time.Now(),
+			DID:     did,
+			Err:     nil,
+		}
+	}
```

Comment:
Slightly confusing/worrying due to the error not being handled directly after the call, seems to make sense but could use a comment.

Reply:
good call, this is a footgun. for all the "is the error nil" checks, I flipped to `nil == err` (so it is more visually distinct from the usual `err != nil`), and added a comment, and added a longer comment in this particular "capture err but don't return it directly" situation.